Jenkinsfile

```diff
@@ -1,192 +1,132 @@
-        daysToKeepStr: '',
-        numToKeepStr: ''
-    ]
+  $class: 'BuildDiscarderProperty',
+  strategy: [
+    $class: 'LogRotator',
+    artifactDaysToKeepStr: '',
+    artifactNumToKeepStr: '10',
+    daysToKeepStr: '',
+    numToKeepStr: ''
+  ]
 ]]);
 
-def failure_function(exception_obj, failureMessage) {
-    def toEmails = [[$class: 'DevelopersRecipientProvider']]
-    emailext body: '${DEFAULT_CONTENT}\n\"' + failureMessage + '\"\n\nCheck console output at $BUILD_URL to view the results.', recipientProviders: toEmails, subject: '${DEFAULT_SUBJECT}'
-    slackSend color: 'danger', message: "${project}: " + failureMessage
-    throw exception_obj
-}
-
-def create_container(image_key) {
-    def image = docker.image(images[image_key]['name'])
-    def container = image.run("\
-        --name ${container_name(image_key)} \
-        --tty \
-        --network=host \
-        --env http_proxy=${env.http_proxy} \
-        --env https_proxy=${env.https_proxy} \
-        --env local_conan_server=${env.local_conan_server} \
-          ")
-}
-
-def docker_copy_code(image_key) {
-    def custom_sh = images[image_key]['sh']
-    sh "docker cp ${project} ${container_name(image_key)}:/home/jenkins/${project}"
-    sh """docker exec --user root ${container_name(image_key)} ${custom_sh} -c \"
-                        chown -R jenkins.jenkins /home/jenkins/${project}
-                        \""""
-}
-
-def docker_dependencies(image_key) {
-    try {
-        def custom_sh = images[image_key]['sh']
-        def conan_remote = "ess-dmsc-local"
-        def dependencies_script = """
-                        mkdir build
-                        cd build
-                        conan remote add \
-                            --insert 0 \
-                            ${conan_remote} ${local_conan_server}
-                        conan install --build=outdated ../${project}/
-                    """
-        sh "docker exec ${container_name(image_key)} ${custom_sh} -c \"${dependencies_script}\""
-    } catch (e) {
-        failure_function(e, "Add conan remote for (${container_name(image_key)}) failed")
-    }
-}
+clangformat_os = "fedora25"
+test_os = "ubuntu1804"
 
-def docker_cmake(image_key) {
-    try {
-        def custom_sh = images[image_key]['sh']
-        def configure_script = """
-                        cd build
-                        . ./activate_run.sh
-                        cmake ../${project} -DBUILD_EVERYTHING=ON
-                    """
-        sh "docker exec ${container_name(image_key)} ${custom_sh} -c \"${configure_script}\""
-    } catch (e) {
-        failure_function(e, "CMake step for (${container_name(image_key)}) failed")
-    }
-}
+container_build_nodes = [
+  'centos7': ContainerBuildNode.getDefaultContainerBuildNode('centos7'),
+  'ubuntu1804': ContainerBuildNode.getDefaultContainerBuildNode('ubuntu1804'),
+  'fedora25': new ContainerBuildNode('essdmscdm/fedora25-build-node:2.0.0', 'bash -e')
```

Comment:
Why does line 25 do `new ContainerBuildNode()` whereas the previous ones does not.

Reply:
In lines 23 and 24, a static function that returns the default build node is being used. Internally it uses `new ContainerBuildNode()`. We don't provide a default version of a Fedora build node, hence the constructor is used directly in line 25.